File: src/main/java/ro/sopa/statistifier/db/repository/TrackListenRepository.java
package ro.sopa.statistifier.db.repository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import ro.sopa.statistifier.db.model.TrackListen;

import java.time.ZonedDateTime;
import java.util.List;

@org.springframework.stereotype.Repository
public interface TrackListenRepository extends JpaRepository<TrackListen, Long> {

    @Query("select max(t.date) from TrackListen t where t.username=:username")
    ZonedDateTime findLatestListenDate(@Param("username") String username);

//    @Query(value = "select t.artist, count(1) as c from Track_Listen t where t.username = :username group by t.artist order by c desc limit :number", nativeQuery = true)
    @Query(value = "select top (:number) t.artist, count(1) as c from Track_Listen t where t.username = :username group by t.artist order by c desc", nativeQuery = true)
    List<Object[]> findTopArtists(@Param("username") String username, @Param("number") Integer number);

    @Query("select t.date from TrackListen t where t.username = :username and artist = :artist and date >= :minDate and date <= :maxDate")
    List<ZonedDateTime> findListenDates(@Param("username") String username, @Param("artist") String artist, @Param("minDate")  ZonedDateTime minDate, @Param("maxDate") ZonedDateTime maxDate);

}

File: src/main/kotlin/ro/sopa/lastfm/db/repository/TrackListenRepository.kt
package ro.sopa.lastfm.db.repository

import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import ro.sopa.lastfm.db.ArtistCountDto
import ro.sopa.lastfm.db.model.TrackListen
import java.time.ZonedDateTime

@Repository
interface TrackListenRepository : CrudRepository<TrackListen, Int>{

    @Query("select max(t.date) from TrackListen t where t.userId=:userId")
    fun findLatestListenDate(@Param("userId") userId: Int): ZonedDateTime?

    @Query("select min(t.date) from TrackListen t where t.userId=:userId")
    fun findOldestListenDate(@Param("userId") userId: Int): ZonedDateTime?

    @Query(value = "select t.artist_id as artistId, count(1) as count from track_Listens t where t.user_Id = :userId group by t.artist_id order by count desc limit :number", nativeQuery = true)
    fun findTopArtists_mysql(@Param("userId") userId: Int, @Param("number") number: Int?): List<ArtistCountDto>

    @Query(value = "select t.artist_id as artistId, count(1) as count from track_Listens t where t.user_Id = :userId and t.date >= :from and t.date <= :to group by t.artist_id order by count desc limit :number", nativeQuery = true)
    fun findTopArtists_mysql(@Param("userId") userId: Int, @Param("number") number: Int?, @Param("from") from: ZonedDateTime, @Param("to") to: ZonedDateTime): List<ArtistCountDto>

    @Query(value = "select top (:number) t.artist_id as artistId, count(1) as count from track_Listens t where t.user_Id = :userId group by t.artist_id order by count desc", nativeQuery = true)
    fun findTopArtists_mssql(@Param("userId") userId: Int, @Param("number") number: Int?): List<ArtistCountDto>

    @Query(value = "select top (:number) t.artist_id as artistId, count(1) as count from track_Listens t where t.user_Id = :userId and t.date >= :from and t.date <= :to group by t.artist_id order by count desc", nativeQuery = true)
    fun findTopArtists_mssql(@Param("userId") userId: Int, @Param("number") number: Int?, @Param("from") from: ZonedDateTime, @Param("to") to: ZonedDateTime): List<ArtistCountDto>

    @Query("select t.date from TrackListen t where t.userId = :userId and artistId = :artistId and date >= :minDate and date <= :maxDate")
    fun findListenDates(
        @Param("userId") userId: Int,
        @Param("artistId") artist: Int,
        @Param("minDate") minDate: ZonedDateTime?,
        @Param("maxDate") maxDate: ZonedDateTime?
    ): List<ZonedDateTime>

}

object TrackListenRepositoryExtensions {

    fun TrackListenRepository.findTopArtists(userId: Int, number: Int, dbType: String): List<ArtistCountDto> {
        if ("mysql" == dbType) {
            return findTopArtists_mysql(userId, number)
        }
        if ("mssql" == dbType) {
            return findTopArtists_mssql(userId, number)
        }
        throw RuntimeException("Unknown db type: $dbType")
    }

    fun TrackListenRepository.findTopArtists(userId: Int, number: Int, dbType: String, from: ZonedDateTime, to: ZonedDateTime): List<ArtistCountDto> {
        if ("mysql" == dbType) {
            return findTopArtists_mysql(userId, number, from, to)
        }
        if ("mssql" == dbType) {
            return findTopArtists_mssql(userId, number, from, to)
        }
        throw RuntimeException("Unknown db type: $dbType")
    }

}